pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -0,0 +1,110 @@
+     */
+    private fun visit(node: Node?, parent: ApexNode<*>?): AbstractApexNode<*>? {
+        val wrapper: AbstractApexNode<*>? =
+            when (node) {
+                null -> null
+                is CompilationUnit -> visit(node.typeDeclaration, parent)
+                is TypeDeclaration -> visitTypeDeclaration(node)
+                is Identifier,
+                is KeywordModifier,
+                is TypeRef -> null
+                else -> {
+                    println("No adapter exists for type ${node::class.qualifiedName}")
+                    // TODO(b/239648780): temporary print
+                    null
+                }
+            }
+
+        wrapper?.setParent(parent)
+        wrapper?.handleSourceCode(sourceCode)
+        wrapper?.calculateLineNumbers(sourceCodePositioner)
+        return wrapper
+    }
+
+    /** Calls [visit] on each of [nodes]. */
+    private fun visit(nodes: List<Node>, parent: ApexNode<*>?) = nodes.forEach { visit(it, parent) }
+
+    /** Calls [visit] on each [child][Node.getChildren] of [node]. */
+    private fun visitAll(node: Node, parent: ApexNode<*>?) =
+        node.getChildren().forEach { visit(it, parent) }
+
+    /**
+     * Calls [visit] on each [child][Node.getChildren] of [node] that isn't already [visited].
+     *
+     * [visited] is a predicate that, given a child node `child`, should return `true` when `child` is
+     * equal to a property of [node] that was already visited and shouldn't be visited again.
+     */
+    private fun visitRest(node: Node, parent: ApexNode<*>?, visited: (Node) -> Boolean) =
+        node.getChildren().filterNot(visited).forEach { visit(it, parent) }
+
+    /** Visits the [TypeDeclaration] node and returns an [ApexRootNode] wrapper. */
+    private fun visitTypeDeclaration(node: TypeDeclaration) =
+        when (node) {
+            is ClassDeclaration ->
+                ASTUserClass(node).apply {
+                    visitModifiers(node.modifiers, parent = this)
+                    visitRest(node, parent = this, visited = { it in node.modifiers })
+                }
+            is InterfaceDeclaration ->
+                ASTUserInterface(node).apply {
+                    visitModifiers(node.modifiers, parent = this)
+                    visitRest(node, parent = this, visited = { it in node.modifiers })
+                }
+            is EnumDeclaration -> ASTUserEnum(node) // TODO(b/239648780): enum body is untranslated
+            is TriggerDeclaration -> ASTUserTrigger(node) // TODO(b/239648780): visit children
+        }
+
+    /**
+     * Visits a [List] of [Modifier]s and adds the resulting [ASTModifierNode] wrapper to [parent].
+     */
+    private fun visitModifiers(modifiers: List<Modifier>, parent: ApexNode<*>) =
+        ASTModifierNode(ListAdapter(modifiers, loc = SourceLocation.UNKNOWN)).apply {
+            visit(modifiers, parent = this)
+            this.setParent(parent)
```

Comment:
Does the `ASTModifierNode` need the `.handleSourceCode` and `.calculateLineNumbers` called on it?

Reply:
Thank you for pointing this out. In the original tree builder, since all nodes are handled equally, there was a single place that called these functions on all nodes being built. It's important that all nodes that end up in the tree receive these calls. 

I believe adding these calls as a separate pass after the tree is completely built would also work. Since we are creating nodes all over the file, this way we can keep the calls in one place too. I'll try making this change in a Jorje version first so that unit tests can confirm whether it's equivalent.